File: 24-kth-smallest/solution.py
# Kth Smallest Element in Array
# Time: O(n) average with quickselect, O(n log k) with heap
# Space: O(1) quickselect / O(k) heap

from typing import List
import heapq
import random

class Solution:
    # Using min heap - O(n log k)
    def findKthSmallestHeap(self, nums: List[int], k: int) -> int:
        return heapq.nsmallest(k, nums)[-1]
    
    # Using quickselect - O(n) average
    def findKthSmallest(self, nums: List[int], k: int) -> int:
        def quickselect(left, right, k_smallest):
            if left == right:
                return nums[left]
            
            pivot_idx = random.randint(left, right)
            pivot_idx = partition(left, right, pivot_idx)
            
            if k_smallest == pivot_idx:
                return nums[k_smallest]
            elif k_smallest < pivot_idx:
                return quickselect(left, pivot_idx - 1, k_smallest)
            else:
                return quickselect(pivot_idx + 1, right, k_smallest)
        
        def partition(left, right, pivot_idx):
            pivot = nums[pivot_idx]
            nums[pivot_idx], nums[right] = nums[right], nums[pivot_idx]
            store_idx = left
            
            for i in range(left, right):
                if nums[i] < pivot:
                    nums[store_idx], nums[i] = nums[i], nums[store_idx]
                    store_idx += 1
            
            nums[right], nums[store_idx] = nums[store_idx], nums[right]
            return store_idx
        
        return quickselect(0, len(nums) - 1, k - 1)

# Test
if __name__ == "__main__":
    sol = Solution()
    print(sol.findKthSmallest([3, 2, 1, 5, 6, 4], 2))  # 2
    print(sol.findKthSmallestHeap([3, 2, 3, 1, 2, 4, 5, 5, 6], 4))  # 3

File: 24-kth-smallest/solution.js
// Kth Smallest Element in Array
// Time: O(n) average with quickselect, O(n log k) with heap
// Space: O(1) quickselect / O(k) heap

class Solution {
    // Using quickselect - O(n) average
    findKthSmallest(nums, k) {
        const quickselect = (left, right, kSmallest) => {
            if (left === right) return nums[left];
            
            const pivotIdx = left + Math.floor(Math.random() * (right - left + 1));
            const newPivotIdx = partition(left, right, pivotIdx);
            
            if (kSmallest === newPivotIdx) {
                return nums[kSmallest];
            } else if (kSmallest < newPivotIdx) {
                return quickselect(left, newPivotIdx - 1, kSmallest);
            } else {
                return quickselect(newPivotIdx + 1, right, kSmallest);
            }
        };
        
        const partition = (left, right, pivotIdx) => {
            const pivot = nums[pivotIdx];
            [nums[pivotIdx], nums[right]] = [nums[right], nums[pivotIdx]];
            let storeIdx = left;
            
            for (let i = left; i < right; i++) {
                if (nums[i] < pivot) {
                    [nums[storeIdx], nums[i]] = [nums[i], nums[storeIdx]];
                    storeIdx++;
                }
            }
            
            [nums[right], nums[storeIdx]] = [nums[storeIdx], nums[right]];
            return storeIdx;
        };
        
        return quickselect(0, nums.length - 1, k - 1);
    }
    
    // Using sort - O(n log n)
    findKthSmallestSort(nums, k) {
        nums.sort((a, b) => a - b);
        return nums[k - 1];
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.findKthSmallest([3, 2, 1, 5, 6, 4], 2));  // 2
    console.log(sol.findKthSmallestSort([3, 2, 3, 1, 2, 4, 5, 5, 6], 4));  // 3
}

module.exports = Solution;
